packages/iris-grid/src/IrisGridMetricCalculator.ts

```diff
@@ -1,31 +1,101 @@
-  >;
-  quickFilters: Map<
-    string,
-    { text: string; filter: dh.FilterCondition | null }
-  >;
-  sorts: dh.Sort[];
+  advancedFilters: ReadonlyAdvancedFilterMap;
+  quickFilters: ReadonlyQuickFilterMap;
+  sorts: readonly dh.Sort[];
   reverse: boolean;
 }
 
-/**
- * Class to calculate all the metrics for a grid.
- * Call getMetrics() with the state to get metrics
- */
 export class IrisGridMetricCalculator extends GridMetricCalculator {
+  // Column widths by name to keep track of columns going in and out of viewport
+  userColumnWidthsByName: Map<ColumnName, number> = new Map();
+
+  // Cached model column names to detect when the column width map update is necessary
+  private cachedModelColumnNames: readonly ColumnName[] | undefined;
+
+  private getCachedCurrentModelColumnNames = memoizeOne(
+    (columns: readonly dh.Column[]) => columns.map(col => col.name)
+  );
+
+  private updateCalculatedColumnWidths(model: IrisGridModel): void {
+    assertNotNull(this.cachedModelColumnNames);
+    const calculatedColumnWidthsByName = new Map<ColumnName, number>();
+    this.cachedModelColumnNames.forEach((name, index) => {
+      const prevColumnWidth = this.calculatedColumnWidths.get(index);
+      if (prevColumnWidth != null) {
+        calculatedColumnWidthsByName.set(name, prevColumnWidth);
+      }
+    });
+    this.resetCalculatedColumnWidths();
+    calculatedColumnWidthsByName.forEach((width, name) => {
+      const index = model.getColumnIndexByName(name);
+      if (index != null) {
+        this.calculatedColumnWidths.set(index, width);
+      }
+    });
+    trimMap(this.calculatedColumnWidths);
+  }
+
+  /**
+   * Updates the user column widths based on the current model state
+   * @param model The current IrisGridModel
+   */
+  private updateUserColumnWidths(model: IrisGridModel): void {
+    this.userColumnWidths = new Map<ModelIndex, number>();
+    this.userColumnWidthsByName.forEach((width, name) => {
+      const modelIndex = model.getColumnIndexByName(name);
+      if (modelIndex != null) {
+        super.setColumnWidth(modelIndex, width);
+      }
+    });
+  }
+
+  /**
+   * Updates the user and calculated column widths if the model columns have changed
+   * @param model The current IrisGridModel
+   */
+  private updateColumnWidthsIfNecessary(model: IrisGridModel): void {
+    // Comparing model.columns references wouldn't work here because
+    // the reference can change in the model without the actual column definitions changing
+    if (
+      this.cachedModelColumnNames != null &&
+      !deepEqual(
+        this.getCachedCurrentModelColumnNames(model.columns),
+        this.cachedModelColumnNames
+      )
+    ) {
+      // Preserve column widths when possible to minimize visual shifts in the grid layout
+      this.updateCalculatedColumnWidths(model);
+      this.updateUserColumnWidths(model);
+    }
+    this.cachedModelColumnNames = model.columns.map(col => col.name);
```

Comment:
This could just use the result from `this.getCachedCurrentModelColumnNames`... I think I'd just rewrite this a bit:
```typescript
const modelColumnNames = this.getCachedCurrentModelColumnNames(model.columns);
if (this.cachedModelColumnNames != null && this.cachedModelColumnNames !== modelColumnNames && !deepEqual(modelColumnNames, cachedModelColumnNames) {
  ...
}
this.cachedModelColumnNames = modelColumnNames;
```

Then you're only doing the `model.columns.map` at most once when this is called (instead of twice), and you can also skip doing a `deepEqual` if the columns haven't actually changed.